pkg/notifications/notifications.go

```diff
@@ -0,0 +1,428 @@
+// entity. Cross-table column references are not supported unless the provided entity provides the fields in one way
+// or another.
+//
+// This allows a query like the following:
+//
+//	> select * from host where id = :host_id and environment_id = :environment_id and name like 'prefix_%'
+//
+// The :host_id and :environment_id parameters will be bound to the entity's ID and EnvironmentId fields, respectively.
+func (client *Client) evaluateRulesForObject(ctx context.Context, entity database.Entity) ([]string, error) {
+	outRuleIds := make([]string, 0, len(client.rulesInfo.Rules))
+
+	for id, filterExpr := range client.rulesInfo.Rules {
+		if filterExpr == "" {
+			outRuleIds = append(outRuleIds, id)
+			continue
+		}
+
+		evaluates, err := func() (bool, error) {
+			// The raw SQL query in the database is URL-encoded (mostly the space character is replaced by %20).
+			// So, we need to unescape it before passing it to the database.
+			query, err := url.QueryUnescape(filterExpr)
+			if err != nil {
+				return false, errors.Wrapf(err, "cannot unescape rule %q object filter expression %q", id, filterExpr)
+			}
+			rows, err := client.db.NamedQueryContext(ctx, client.db.Rebind(query), entity)
+			if err != nil {
+				return false, err
+			}
+			defer func() { _ = rows.Close() }()
+
+			if !rows.Next() {
+				return false, nil
+			}
+			return true, nil
+		}()
+		if err != nil {
+			return nil, errors.Wrapf(err, "cannot fetch rule %q from %q", id, filterExpr)
+		} else if !evaluates {
+			continue
+		}
+		outRuleIds = append(outRuleIds, id)
+	}
+
+	return outRuleIds, nil
+}
+
+// buildCommonEvent creates an event.Event based on Host and (optional) Service names.
+//
+// This function is used by all event builders to create a common event structure that includes the host and service
+// names, the absolute URL to the Icinga Web 2 Icinga DB page for the host or service, and the tags for the event.
+// Any event type-specific information (like severity, message, etc.) is added by the specific event builders.
+func (client *Client) buildCommonEvent(rlr *redisLookupResult) (*event.Event, error) {
+	var (
+		objectName string
+		objectUrl  url.URL
+		objectTags map[string]string
+	)
+
+	if rlr.ServiceName != "" {
+		objectName = rlr.HostName + "!" + rlr.ServiceName
+
+		objectUrl.Path = "/icingadb/service"
+		objectUrl.RawQuery = "name=" + utils.RawUrlEncode(rlr.ServiceName) + "&host.name=" + utils.RawUrlEncode(rlr.HostName)
+
+		objectTags = map[string]string{
+			"host":    rlr.HostName,
+			"service": rlr.ServiceName,
+		}
+	} else {
+		objectName = rlr.HostName
+
+		objectUrl.Path = "/icingadb/host"
+		objectUrl.RawQuery = "name=" + utils.RawUrlEncode(rlr.HostName)
+
+		objectTags = map[string]string{
+			"host": rlr.HostName,
+		}
+	}
+
+	return &event.Event{
+		Name: objectName,
+		URL:  objectUrl.String(),
+		Tags: objectTags,
+	}, nil
+}
```

Comment:
I don't think `url.URL` is still doing anything useful anymore, it just inserts a `?` now. That could also be string concatenation.

Also, the variables provide little benefit here, both branches could just `return &event.Event{...}` directly.

Finally, there's no more case returning an error, so that could be removed from the function signature.

Reply:
Changed to string and reworked method a bit.